File: messages/recover.py
import json
import sys

def apply_replacements(file_path, transcript_path):
    # Read original file
    with open(file_path, 'r') as f:
        content = f.read()
    lines = content.split('\n')
    
    # Parse transcript
    with open(transcript_path, 'r') as f:
        for line in f:
            try:
                entry = json.loads(line)
                if entry.get("type") == "PLANNER_RESPONSE":
                    calls = entry.get("tool_calls", [])
                    for call in calls:
                        if call.get("name") in ("multi_replace_file_content", "replace_file_content"):
                            args = call.get("args", {})
                            if args.get("TargetFile", "").endswith("fi.json"):
                                chunks = args.get("ReplacementChunks", [])
                                if call.get("name") == "replace_file_content":
                                    chunks = [args]
                                
                                # Sort chunks by StartLine in descending order to avoid line number shifts
                                chunks.sort(key=lambda x: x.get("StartLine", 0), reverse=True)
                                
                                for chunk in chunks:
                                    start = chunk.get("StartLine")
                                    end = chunk.get("EndLine")
                                    replacement = chunk.get("ReplacementContent", "")
                                    
                                    if start and end:
                                        # Replace lines from start to end (1-indexed)
                                        prefix = lines[:start-1]
                                        suffix = lines[end:]
                                        lines = prefix + replacement.split('\n') + suffix
            except Exception as e:
                pass

    with open(file_path, 'w') as f:
        f.write('\n'.join(lines))

apply_replacements("fi.json", "/Users/umurey/.gemini/antigravity/brain/ddff770d-ccf3-4e1d-8e79-5382d2464de8/.system_generated/logs/transcript_full.jsonl")

File: messages/recover.js
const fs = require('fs');

const logPath = '/Users/umurey/.gemini/antigravity/brain/d0897114-d182-4b18-a066-5892d8a60c80/.system_generated/logs/transcript_full.jsonl';
const lines = fs.readFileSync(logPath, 'utf8').split('\n').filter(Boolean);

let ka = fs.readFileSync('ka.json', 'utf8');

let applied = 0;
let skipped = 0;

for (const line of lines) {
    try {
        const obj = JSON.parse(line);
        if (obj.tool_calls) {
            for (const call of obj.tool_calls) {
                if (call.name === 'replace_file_content' && call.args.TargetFile && call.args.TargetFile.endsWith('ka.json')) {
                    const target = call.args.TargetContent;
                    const replacement = call.args.ReplacementContent;
                    
                    if (ka.includes(target)) {
                        ka = ka.replace(target, replacement);
                        applied++;
                    } else {
                        skipped++;
                    }
                }
                else if (call.name === 'multi_replace_file_content' && call.args.TargetFile && call.args.TargetFile.endsWith('ka.json')) {
                    if (call.args.ReplacementChunks) {
                        for (const chunk of call.args.ReplacementChunks) {
                            if (ka.includes(chunk.TargetContent)) {
                                ka = ka.replace(chunk.TargetContent, chunk.ReplacementContent);
                                applied++;
                            } else {
                                skipped++;
                            }
                        }
                    }
                }
            }
        }
    } catch(e) {}
}

console.log(`Applied ${applied} edits. Skipped ${skipped} edits.`);
fs.writeFileSync('ka_recovered_full.json', ka);
